backend/src/main/scala/io/udash/web/guide/markdown/MarkdownPagesEndpoint.scala

```diff
@@ -0,0 +1,35 @@
+package io.udash.web.guide.markdown
+
+import java.io.{BufferedReader, File, FileReader}
+import java.time.Instant
+import java.util.concurrent.ConcurrentHashMap
+
+import com.avsystem.commons._
+import com.vladsch.flexmark.html.HtmlRenderer
+import com.vladsch.flexmark.parser.Parser
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class MarkdownPagesEndpoint(guideResourceBase: String)(implicit ec: ExecutionContext) extends MarkdownPageRPC {
+
+  private val parser = Parser.builder.build
+  private val renderer = HtmlRenderer.builder.build
+  private val renderedPages = new ConcurrentHashMap[MarkdownPage, (Future[String], Instant)]
```

Comment:
shouldn't we store rendered `String` here? not a future?

Reply:
We wouldn't be able to distinguish between the "failed with error" and "never rendered" cases and we would have to smh mark ongoing computations to avoid multiple renderings for the same file when client(s) request N in a row.